src/ws/session.c

```diff
@@ -327,30 +366,55 @@ pam_conv_func (int num_msg,
 
   for (i = 0; i < num_msg; i++)
     {
-      if (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF)
+      if (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF &&
+          *password != NULL)
+        {
+            debug ("answered pam password prompt");
+            resp[i].resp = *password;
+            resp[i].resp_retcode = 0;
+            *password = NULL;
+        }
+      else if (msg[i]->msg_style == PAM_ERROR_MSG)
         {
-          if (*password == NULL)
+          if (asprintf (&err_msg, "%s ", msg[i]->msg) < 0)
             {
-              warnx ("pam asked us for unexpected password");
+              warnx ("Error allocating error variable");
               success = 0;
-            }
-          else
-            {
-              debug ("answered pam password prompt");
-              resp[i].resp = *password;
-              resp[i].resp_retcode = 0;
-              *password = NULL;
+              break;
             }
         }
-      else if (msg[i]->msg_style == PAM_ERROR_MSG ||
-               msg[i]->msg_style == PAM_TEXT_INFO)
+      else if (msg[i]->msg_style == PAM_TEXT_INFO)
         {
-          warnx ("pam: %s", msg[i]->msg);
+          if (asprintf (&txt_msg, "%s ", msg[i]->msg) < 0)
+            {
+              warnx ("Error allocating text variable");
+              success = 0;
+              break;
+            }
         }
       else
         {
-          warnx ("pam asked us for an unsupported info: %s", msg[i]->msg);
-          success = 0;
+          debug ("prompt for more data");
+          write_auth_begin ();
+          if (txt_msg)
+            write_auth_string ("message", txt_msg);
+          if (err_msg)
+            write_auth_string ("error", err_msg);
```

Comment:
Right now I have login html showing a password input for all prompts. Do you think we need to switch between text and password inputs based on echo?